File: api/src/main/java/com/mengyunzhi/springBootStudy/service/CourseService.java
package com.mengyunzhi.springBootStudy.service;


import com.mengyunzhi.springBootStudy.entity.Course;
import com.mengyunzhi.springBootStudy.entity.Term;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;

import javax.validation.constraints.NotNull;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * 课程
 * @author panjie
 */
public interface CourseService {
    /**
     * 新增课程
     * @param course 课程
     * @return 课程
     */
    ResponseEntity<Map<String, Object>> save(Course course);

    /**
     * 名称是否存在
     * @param name 课程名称
     * @return true 存在
     */
    boolean existsByName(String name);

    /**
     * 删除
     *
     * @param id 课程ID
     */
    void deleteById(Long id);

    /**
     * 获取所有课程列表
     *
     * @param name 课程名称
     * @return
     */
    List<Course> getAll(String name);

    /**
     * 通过ID获取课程
     *
     * @param id 课程ID
     * @return 课程实体
     */
    Course getById(Long id);

    /**
     * 更新课程
     *
     * @param id    预更新的课程ID
     * @param klass 新的课程信息
     */
    ResponseEntity<Map<String, Object>> update(Long id, Course klass);

    /**
     * 综合查询
     * @param name 课程名
     * @param klassId 班级id
     * @param schoolId 学校id
     * @param termId 学期id
     * @param userId 学生id
     * @param pageable 分页参数
     * */
    Page<Course> findAll(String name, Long schoolId, Long klassId, Long termId, Long userId, @NotNull Pageable pageable);

    Page<Course> findCoursesByCriteria(Long termId, String courseName, Long sory, Long userId, @NotNull Pageable pageable);

    ArrayList<Course> getCoursesByTermId(Long termId, Long sory);

    ArrayList<Course> findByTermId(Long termId);
    /**
     * 通过id寻找course
     *
     * @param courseId 课程id
     **/
    Optional<Course> findById(Long courseId);

    List<Course> findAll(Long schoolId, Long klassId, Long termId, Long usrId, List<Integer> week);
}

File: web-app/src/app/service/course.service.ts
import {Injectable} from '@angular/core';
import {Course} from '../norm/entity/Course';
import {Observable} from 'rxjs';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Page} from '../norm/entity/page';
import { User } from '../norm/entity/User';
import {Term} from "../norm/entity/Term";

@Injectable({
  providedIn: 'root'
})
export class CourseService {
  private url = 'http://localhost:8080/Course';

  constructor(private httpClient: HttpClient) {
  }

  /**
   * 保存课程
   * @param course 课程
   */
  save(course: Course | undefined): Observable<Course> {
    return this.httpClient.post<Course>(this.url, course);
  }

  /**
   * 课程名称是否存在
   * @param name 课程名称
   */
  existsByName(name: string): Observable<boolean> {
    const url = this.url + '/existsByName';
    return this.httpClient.get<boolean>(url, {params: {name}});
  }

  /**
   * 分页
   * @param params name课程名称  klassId 班级 teacherId 教师
   */
  page(params: {
    schoolId?: number,
    klassId?: number,
    termId?:  number,
    userId?: number,
    name?: string,
    page: number,
    size: number
  }): Observable<Page<Course>> {
     /* 设置默认值*/
     if (params.page === undefined) {
      params.page = 0;
    }
    if (params.size === undefined) {
      params.size = 10;
    }

    /* 初始化查询参数 */
    const queryParams = new HttpParams()
      .set('name', params.name ? params.name : '')
      .set('schoolId', params.schoolId ? params.schoolId.toString() : '')
      .set('klassId', params.klassId ? params.klassId.toString() : '')
      .set('termId', params.termId ? params.termId.toString() : '')
      .set('userId', params.userId ? params.userId.toString() : '')
      .set('page', params.page.toString())
      .set('size', params.size.toString());
    return this.httpClient.get<Page<Course>>(this.url, {params: queryParams});
  }

  /**
   * 增加新课程
   *
   *
   ***/
  add(course:  {
    name?: string | null,
    sory?: number | null,
    week?: (number | null)[],
    day?: number  | null,
    period?: number | null,
    schoolId?: number | null,
    clazz_id?: number | null,
    term_id?: number | null,
    userId?: number | null
  }): Observable<void> {

    let newCourse = {
      name: course.name,
      sory: course.sory,
      week: course.week,
      day: [course.day],
      period: [course.day],
      schoolId: course.schoolId,
      klassId: course.clazz_id,
      termId: course.term_id,
      userId: course.userId
    }
    return this.httpClient.post<void>(this.url + '/add', newCourse);
  }

  /**
   * 获取课程通过id
   * @param id 课程id
   * **/
  getById(courseId: number): Observable<Course> {
    /* 初始化查询参数 */
    const queryParams = new HttpParams()
    .set('courseId', courseId ? courseId.toString() : '')
    return this.httpClient.get<Course>(this.url + '/findById', {params: queryParams})
  }

  /**
   * 更新课程
   * @param course 更新的信息
   * **/
  update(course:  {
    id: number
    name?: string | null,
    sory?: number | null,
    week?: (number | null)[],
    day?: number  | null,
    period?: number | null,
    schoolId?: number | null,
    clazz_id?: number | null,
    term_id?: number | null,
    userId?: number | null
  }): Observable<void> {
    let updateCourse = {
      id: course.id,
      name: course.name,
      sory: course.sory,
      week: course.week,
      day: [course.day],
      period: [course.day],
      schoolId: course.schoolId,
      klassId: course.clazz_id,
      termId: course.term_id,
      userId: course.userId
    }
    return this.httpClient.put<void>(this.url + '/updateCourse', updateCourse)
  }

  /**
   * 删除对应课程
   * @param courseId 课程id
   */
  onDelete(courseId: number): Observable<void> {
    return this.httpClient.delete<void>(this.url + '/delete/' + courseId);
  }
}
